[Bug]: GHC cannot use base modules libraries

### Problem description

GHC cannot compile anything because it asks for normal interface files `.hi` where's only dynamic linking interface files `.dyn_hi` are available in the base package. Strangely enough, the interactive interface CGHi works just fine.

also just a heads up i've already checked `ghc-pkg check` and it only complains about haddock files

### What steps will reproduce the bug?

```log
~ $ cat test.hs
main = putStrLn "test"
~ $ ghc -v test.hs
Glasgow Haskell Compiler, Version 8.10.7, stage 2 booted by GHC version 8.10.7
*** initializing package database:
Using binary package database: /data/data/com.termux/files/usr/lib/ghc-8.10.7/package.conf.d/package.cache
package flags []
loading package database /data/data/com.termux/files/usr/lib/ghc-8.10.7/package.conf.d
wired-in package ghc-prim mapped to ghc-prim-0.6.1
wired-in package integer-wired-in mapped to integer-gmp-1.0.3.0
wired-in package base mapped to base-4.14.3.0
wired-in package rts mapped to rts
wired-in package template-haskell mapped to template-haskell-2.16.0.0
wired-in package ghc mapped to ghc-8.10.7
!!! initializing package database: finished in 103.19 milliseconds, allocated 8.495 megabytes
*** initializing package database:
package flags []
loading package database /data/data/com.termux/files/usr/lib/ghc-8.10.7/package.conf.d
wired-in package ghc-prim mapped to ghc-prim-0.6.1
wired-in package integer-wired-in mapped to integer-gmp-1.0.3.0
wired-in package base mapped to base-4.14.3.0
wired-in package rts mapped to rts-1.0.1
wired-in package template-haskell mapped to template-haskell-2.16.0.0
wired-in package ghc mapped to ghc-8.10.7
!!! initializing package database: finished in 69.92 milliseconds, allocated 3.244 megabytes
*** Chasing dependencies:
Chasing modules from: *test.hs
!!! Chasing dependencies: finished in 1.97 milliseconds, allocated 0.211 megabytes
Stable obj: {}
Stable BCO: {}
Ready for upsweep
  [NONREC
      ModSummary {
         ms_hs_date = 2022-11-03 14:26:03.850077978 UTC
         ms_mod = Main,
         ms_textual_imps = [(Nothing, Prelude)]
         ms_srcimps = []
      }]
*** Deleting temp files:
Deleting:
compile: input file test.hs
*** Checking old interface for Main (use -ddump-hi-diffs for more details):
[1 of 1] Compiling Main             ( test.hs, test.o )
*** Parser [Main]:
!!! Parser [Main]: finished in 0.68 milliseconds, allocated 0.069 megabytes
*** Renamer/typechecker [Main]:
!!! Renamer/typechecker [Main]: finished in 0.10 milliseconds, allocated 0.022 megabytes

test.hs:1:1: error:
    Could not find module ‘Prelude’
    There are files missing in the ‘base-4.14.3.0’ package,
    try running 'ghc-pkg check'.
    Locations searched:
      Prelude.hs
      Prelude.lhs
      Prelude.hsig
      Prelude.lhsig
      /data/data/com.termux/files/usr/lib/ghc-8.10.7/base-4.14.3.0/Prelude.hi
  |
1 | main = putStrLn "test"
  | ^
Upsweep partially successful.
*** Deleting temp files:
Deleting:
link(batch): upsweep (partially) failed OR
   Main.main not exported; not linking.
*** Deleting temp files:
Deleting:
*** Deleting temp dirs:
Deleting:
~ $ ghc-pkg check
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/unix-2.7.2.2/unix.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/exceptions-0.10.4/exceptions.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/containers-0.6.5.1/containers.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/filepath-1.4.2.1/filepath.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/deepseq-1.4.4.0/deepseq.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/mtl-2.2.2/mtl.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/template-haskell-2.16.0.0/template-haskell.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/Cabal-3.2.1.0/Cabal.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/ghc-heap-8.10.7/ghc-heap.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/time-1.9.3/time.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/directory-1.3.6.0/directory.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/ghc-boot-th-8.10.7/ghc-boot-th.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/integer-gmp-1.0.3.0/integer-gmp.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/pretty-1.1.3.6/pretty.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/ghc-boot-8.10.7/ghc-boot.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/ghci-8.10.7/ghci.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/haskeline-0.8.2/haskeline.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/ghc-prim-0.6.1/ghc-prim.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/ghc-8.10.7/ghc.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/libiserv-8.10.7/libiserv.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/ghc-compact-0.1.0.0/ghc-compact.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/hpc-0.6.1.0/hpc.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/stm-2.5.0.1/stm.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/terminfo-0.4.1.4/terminfo.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/bytestring-0.10.12.0/bytestring.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/process-1.6.13.2/process.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/array-0.5.4.0/array.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/transformers-0.5.6.2/transformers.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/text-1.2.4.1/text.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/base-4.14.3.0/base.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/parsec-3.1.14.0/parsec.haddock doesn't exist or isn't a file
Warning: haddock-interfaces: /data/data/com.termux/files/usr/share/doc/ghc-8.10.7/html/libraries/binary-0.8.8.0/binary.haddock doesn't exist or isn't a file
~ $ ghci
GHCi, version 8.10.7: https://www.haskell.org/ghc/  :? for help
Prelude> putStrLn "test"
test
Prelude> :q
Leaving GHCi.
~ $ ls ../usr/lib/ghc-8.10.7/base-4.14.3.0/                Control         Prelude.dyn_hi
Data            System
Debug           Text
Foreign         Type
Foreign.dyn_hi  Unsafe
GHC             include
Numeric         libHSbase-4.14.3.0-ghc8.10.7.so
Numeric.dyn_hi
```

### What is the expected behavior?

the compilation of a binary executable called `test` that prints `test` on stdout

### System information

termux-info:

```
~ $ termux-info
Termux Variables:
TERMUX_API_VERSION=0.50.1
TERMUX_APK_RELEASE=F_DROID
TERMUX_APP_PACKAGE_MANAGER=apt
TERMUX_APP_PID=18539
TERMUX_IS_DEBUGGABLE_BUILD=0
TERMUX_MAIN_PACKAGE_FORMAT=debian
TERMUX_VERSION=0.118.0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://mirror.vern.cc/termux/termux-main stable main
Updatable packages:
All packages up to date
termux-tools version:
1.31.0
Android version:
11
Kernel build information:
Linux localhost 4.14.186+ #1 SMP PREEMPT Mon Jun 27 01:14:57 CDT 2022 aarch64 Android
Device manufacturer:
motorola
Device model:
motorola edge 20 lite
```

edit: runghc works fine too

--- a/packages/ghc-libs/always-use-pic-on-android-for-dynexec.patch
+++ b/packages/ghc-libs/always-use-pic-on-android-for-dynexec.patch
@@ -1,17 +1,15 @@
-diff --git a/compiler/main/DynFlags.hs b/compiler/main/DynFlags.hs
-index d222b3b..95a9942 100644
---- a/compiler/main/DynFlags.hs
-+++ b/compiler/main/DynFlags.hs
-@@ -4662,6 +4662,12 @@ default_PIC platform =
+--- ghc.orig/compiler/GHC/Driver/Session.hs	2022-11-07 01:10:29.000000000 +0530
++++ ghc.mod/compiler/GHC/Driver/Session.hs	2022-11-10 13:12:21.825682472 +0530
+@@ -3763,6 +3763,12 @@
      (OSDarwin,  ArchAArch64) -> [Opt_PIC]
      (OSLinux,   ArchAArch64) -> [Opt_PIC, Opt_ExternalDynamicRefs]
      (OSLinux,   ArchARM {})  -> [Opt_PIC, Opt_ExternalDynamicRefs]
-+    (OSLinux, ArchX86) -> [Opt_PIC, Opt_ExternalDynamicRefs]
-+    (OSLinux, ArchX86_64) -> [Opt_PIC, Opt_ExternalDynamicRefs]
-+    (OSUnknown, ArchARM {}) -> [Opt_PIC, Opt_ExternalDynamicRefs]
-+    (OSUnknown, ArchAArch64) -> [Opt_PIC, Opt_ExternalDynamicRefs]
-+    (OSUnknown, ArchX86) -> [Opt_PIC, Opt_ExternalDynamicRefs]
-+    (OSUnknown, ArchX86_64) -> [Opt_PIC, Opt_ExternalDynamicRefs]
++    (OSLinux,   ArchX86)     -> [Opt_PIC, Opt_ExternalDynamicRefs]
++    (OSLinux,   ArchX86_64)  -> [Opt_PIC, Opt_ExternalDynamicRefs]
++    (OSUnknown,   ArchAArch64) -> [Opt_PIC, Opt_ExternalDynamicRefs]
++    (OSUnknown,   ArchARM {})  -> [Opt_PIC, Opt_ExternalDynamicRefs]
++    (OSUnknown,   ArchX86)     -> [Opt_PIC, Opt_ExternalDynamicRefs]
++    (OSUnknown,   ArchX86_64)  -> [Opt_PIC, Opt_ExternalDynamicRefs]
      (OSOpenBSD, ArchX86_64)  -> [Opt_PIC] -- Due to PIE support in
-                                           -- OpenBSD since 5.3 release
-                                           -- (1 May 2013) we need to
+                                          -- OpenBSD since 5.3 release
+                                          -- (1 May 2013) we need to

--- a/packages/ghc-libs/always_link_m.patch
+++ b/packages/ghc-libs/always_link_m.patch
@@ -2,7 +2,7 @@
 +++ ghc-8.10.7-patch/rts/package.conf.in	2022-02-02 15:19:26.221569951 +0530
 @@ -26,9 +26,7 @@
  hs-libraries:   "HSrts" FFI_LIB
- 
+
  extra-libraries:
 -#if defined(HAVE_LIBM)
                                  "m"             /* for ldexp() */
@@ -10,9 +10,9 @@
  #if defined(HAVE_LIBRT)
                                , "rt"
  #endif
---- ghc-8.10.7/compiler/main/SysTools.hs	2021-06-29 12:54:51.000000000 +0530
-+++ ghc-8.10.7-patch/compiler/main/SysTools.hs	2022-02-02 15:17:37.251569993 +0530
-@@ -424,11 +424,7 @@
+--- ghc.orig/compiler/GHC/Linker/Dynamic.hs	2022-11-07 01:10:29.000000000 +0530
++++ ghc.mod/compiler/GHC/Linker/Dynamic.hs	2022-11-10 13:19:22.885682312 +0530
+@@ -229,11 +229,7 @@
  -- math-y things are used (which we assume to include all programs). See #14022.
  libmLinkOpts :: [Option]
  libmLinkOpts =
@@ -21,6 +21,6 @@
 -#else
 -  []
 -#endif
- 
- getPkgFrameworkOpts :: DynFlags -> Platform -> [InstalledUnitId] -> IO [String]
- getPkgFrameworkOpts dflags platform dep_packages
+
+ {-
+ Note [-Bsymbolic assumptions by GHC]

--- a/packages/ghc-libs/build.sh
+++ b/packages/ghc-libs/build.sh
@@ -1,12 +1,11 @@
 TERMUX_PKG_HOMEPAGE=https://www.haskell.org/ghc/
-TERMUX_PKG_DESCRIPTION="The Glasgow Haskell Compiler - Dynamic Libraries"
+TERMUX_PKG_DESCRIPTION="The Glasgow Haskell Compiler libraries"
 TERMUX_PKG_LICENSE="BSD 2-Clause, BSD 3-Clause, LGPL-2.1"
 TERMUX_PKG_MAINTAINER="Aditya Alok <alok@termux.org>"
-TERMUX_PKG_VERSION=8.10.7
-TERMUX_PKG_REVISION=3
+TERMUX_PKG_VERSION=9.2.5
 TERMUX_PKG_SRCURL="https://downloads.haskell.org/~ghc/${TERMUX_PKG_VERSION}/ghc-${TERMUX_PKG_VERSION}-src.tar.xz"
-TERMUX_PKG_SHA256=e3eef6229ce9908dfe1ea41436befb0455fefb1932559e860ad4c606b0d03c9d
-TERMUX_PKG_DEPENDS="iconv, libffi, ncurses, libgmp, libandroid-posix-semaphore"
+TERMUX_PKG_SHA256=0606797d1b38e2d88ee2243f38ec6b9a1aa93e9b578e95f0de9a9c0a4144021c
+TERMUX_PKG_DEPENDS="libiconv, libffi, ncurses, libgmp, libandroid-posix-semaphore"
 TERMUX_PKG_BUILD_IN_SRC=true
 TERMUX_PKG_EXTRA_CONFIGURE_ARGS="
 --disable-ld-override
@@ -21,66 +20,33 @@ TERMUX_PKG_EXTRA_CONFIGURE_ARGS="
 --with-iconv-libraries=${TERMUX_PREFIX}/lib
 --with-curses-libraries=${TERMUX_PREFIX}/lib
 --with-curses-includes=${TERMUX_PREFIX}/include
---with-curses-libraries-stage0=/usr/lib
 "
-# ghc-pkg is in this package. Here ghci is lib not bin.
-TERMUX_PKG_PROVIDES="haskekl-ghc-pkg, haskell-ghci"
-TERMUX_PKG_CONFLICTS="haskell-ghci"
-TERMUX_PKG_STATICSPLIT_EXTRA_PATTERNS="lib/**/*.hi lib/**/*.o"
+TERMUX_PKG_NO_STATICSPLIT=true
+TERMUX_PKG_REPLACES="ghc-libs-static"
 
 termux_step_pre_configure() {
 	termux_setup_ghc
 
-	_TERMUX_HOST_PLATFORM="${TERMUX_HOST_PLATFORM}"
-	[ "${TERMUX_ARCH}" = "arm" ] && _TERMUX_HOST_PLATFORM="armv7a-linux-androideabi"
+	local host_platform="${TERMUX_HOST_PLATFORM}"
+	[ "${TERMUX_ARCH}" = "arm" ] && host_platform="armv7a-linux-androideabi"
+	TERMUX_PKG_EXTRA_CONFIGURE_ARGS+=" --target=${host_platform}"
 
-	TERMUX_PKG_EXTRA_CONFIGURE_ARGS+=" --target=${_TERMUX_HOST_PLATFORM}"
-
-	_WRAPPER_BIN="${TERMUX_PKG_BUILDDIR}/_wrapper/bin"
-	mkdir -p "${_WRAPPER_BIN}"
-
-	for tool in llc opt; do
-		local wrapper="${_WRAPPER_BIN}/${tool}"
-		cat >"$wrapper" <<-EOF
-			#!$(command -v sh)
-			exec /usr/lib/llvm-12/bin/${tool} "\$@"
-		EOF
-		chmod 0700 "$wrapper"
-	done
-
-	local ar_wrapper="${_WRAPPER_BIN}/${_TERMUX_HOST_PLATFORM}-ar"
-	cat >"$ar_wrapper" <<-EOF
-		#!$(command -v sh)
-		exec $(command -v ${AR}) "\$@"
-	EOF
-	chmod 0700 "$ar_wrapper"
-
-	local strip_wrapper="${_WRAPPER_BIN}/${_TERMUX_HOST_PLATFORM}-strip"
-	cat >"$strip_wrapper" <<-EOF
-		#!$(command -v sh)
-		exec $(command -v ${STRIP}) "\$@"
-	EOF
-	chmod 0700 "$strip_wrapper"
-
-	export PATH="${_WRAPPER_BIN}:${PATH}"
-	export LIBTOOL="$(command -v libtool)"
-
-	local EXTRA_FLAGS="-O -optl-Wl,-rpath=${TERMUX_PREFIX}/lib -optl-Wl,--enable-new-dtags"
-
-	[ "${TERMUX_ARCH}" != "i686" ] && EXTRA_FLAGS+=" -fllvm"
+	local extra_flags="-O -optl-Wl,-rpath=${TERMUX_PREFIX}/lib -optl-Wl,--enable-new-dtags"
+	[ "${TERMUX_ARCH}" != "i686" ] && extra_flags+=" -fllvm"
 
 	# Suppress warnings for LLVM 13
-	sed -i 's/LlvmMaxVersion=13/LlvmMaxVersion=14/' configure.ac
+	sed -i 's/LlvmMaxVersion=13/LlvmMaxVersion=15/' configure.ac
+
+	export LIBTOOL && LIBTOOL="$(command -v libtool)"
 
 	cp mk/build.mk.sample mk/build.mk
 	cat >>mk/build.mk <<-EOF
 		SRC_HC_OPTS        = -O -H64m
 		GhcStage1HcOpts    = -O
-		GhcStage2HcOpts    = ${EXTRA_FLAGS}
-		GhcLibHcOpts       = ${EXTRA_FLAGS} -optl-landroid-posix-semaphore
+		GhcStage2HcOpts    = ${extra_flags}
+		GhcLibHcOpts       = ${extra_flags} -optl-landroid-posix-semaphore
 		BuildFlavour       = quick-cross
 		GhcLibWays         = v dyn
-		STRIP_CMD          = ${STRIP}
 		BUILD_PROF_LIBS    = NO
 		HADDOCK_DOCS       = NO
 		BUILD_SPHINX_HTML  = NO
@@ -93,21 +59,36 @@ termux_step_pre_configure() {
 	EOF
 
 	patch -p1 <<-EOF
-		--- ghc-8.10.7/rules/build-package-data.mk      2021-06-21 12:24:36.000000000 +0530
-		+++ ghc-8.10.7-patch/rules/build-package-data.mk 2022-01-27 20:31:28.901997265 +0530
+		--- ghc.orig/rules/build-package-data.mk 2022-11-07 01:10:29.000000000 +0530
+		+++ ghc.mod/rules/build-package-data.mk  2022-11-11 13:08:01.992488180 +0530
 		@@ -68,6 +68,12 @@
 		 \$1_\$2_CONFIGURE_LDFLAGS = \$\$(SRC_LD_OPTS) \$\$(\$1_LD_OPTS) \$\$(\$1_\$2_LD_OPTS)
 		 \$1_\$2_CONFIGURE_CPPFLAGS = \$\$(SRC_CPP_OPTS) \$\$(CONF_CPP_OPTS_STAGE\$3) \$\$(\$1_CPP_OPTS) \$\$(\$1_\$2_CPP_OPTS)
 
 		+ifneq "\$3" "0"
-		+ \$1_\$2_CONFIGURE_LDFLAGS += $LDFLAGS
-		+ \$1_\$2_CONFIGURE_CPPFLAGS += $CPPFLAGS
-		+ \$1_\$2_CONFIGURE_CFLAGS += $CFLAGS
+		+ \$1_\$2_CONFIGURE_LDFLAGS += ${LDFLAGS}
+		+ \$1_\$2_CONFIGURE_CPPFLAGS += ${CPPFLAGS}
+		+ \$1_\$2_CONFIGURE_CFLAGS += ${CFLAGS}
 		+endif
 		+
 		 \$1_\$2_CONFIGURE_OPTS += --configure-option=CFLAGS="\$\$(\$1_\$2_CONFIGURE_CFLAGS)"
 		 \$1_\$2_CONFIGURE_OPTS += --configure-option=LDFLAGS="\$\$(\$1_\$2_CONFIGURE_LDFLAGS)"
 		 \$1_\$2_CONFIGURE_OPTS += --configure-option=CPPFLAGS="\$\$(\$1_\$2_CONFIGURE_CPPFLAGS)"
+		@@ -104,9 +110,12 @@
+		 \$1_\$2_CONFIGURE_OPTS += --configure-option=--with-gmp
+		 endif
+
+		-
+		 ifneq "\$\$(CURSES_LIB_DIRS)" ""
+		-\$1_\$2_CONFIGURE_OPTS += --configure-option=--with-curses-libraries="\$\$(CURSES_LIB_DIRS)"
+		+ ifeq "\$3" "0"
+		+  \$1_\$2_CONFIGURE_OPTS += --configure-option=--with-curses-libraries="/usr/lib"
+		+ else
+		+  \$1_\$2_CONFIGURE_OPTS += --configure-option=--with-curses-libraries="\$\$(CURSES_LIB_DIRS)"
+		+ endif
+		 endif
+
+		 \$1_\$2_CONFIGURE_OPTS += --configure-option=--host=\$(TargetPlatformFull)
 	EOF
 
 	./boot

--- a/packages/ghc-libs/correct-host-triplet.patch
+++ b/packages/ghc-libs/correct-host-triplet.patch
@@ -1,6 +1,6 @@
---- ghc-8.10.7/libraries/Cabal/Cabal/Distribution/Simple.hs	2022-02-09 09:22:38.381445950 +0530
-+++ ghc-8.10.7-patch/libraries/Cabal/Cabal/Distribution/Simple.hs	2022-03-05 12:03:02.509066261 +0530
-@@ -105,10 +105,20 @@
+--- ghc.orig/libraries/Cabal/Cabal/src/Distribution/Simple.hs	2022-11-07 01:11:17.000000000 +0530
++++ ghc.mod/libraries/Cabal/Cabal/src/Distribution/Simple.hs	2022-11-10 13:28:41.695682098 +0530
+@@ -108,10 +108,20 @@
  import Distribution.Compat.Environment      (getEnvironment)
  import Distribution.Compat.GetShortPathName (getShortPathName)
 
@@ -11,18 +11,18 @@
 
 +correctHostTriplet :: String -> String
 +correctHostTriplet s = do
-+    if isInfixOf "-android" s
-+      then
-+        let arch = take (length s -8) ( drop 0 s ) -- drop "-android"
-+        in
-+          if arch == "arm" then "armv7a" else arch ++ "-linux-"
-+            ++ if arch == "arm" then "androideabi" else "android"
-+      else s
++  if isInfixOf "-android" s
++    then
++      let arch = take (length s -8) ( drop 0 s ) -- drop "-android"
++      in
++        if arch == "arm" then "armv7a" else arch ++ "-linux-"
++          ++ if arch == "arm" then "androideabi" else "android"
++    else s
 +
  -- | A simple implementation of @main@ for a Cabal setup script.
  -- It reads the package description file using IO, and performs the
  -- action specified on the command line.
-@@ -754,7 +764,7 @@
+@@ -722,7 +732,7 @@
        overEnv = ("CFLAGS", Just cflagsEnv) :
                  [("PATH", Just pathEnv) | not (null extraPath)]
        hp = hostPlatform lbi

--- a/packages/ghc-libs/ghc.subpackage.sh
+++ b/packages/ghc-libs/ghc.subpackage.sh
@@ -1,11 +1,13 @@
+# shellcheck shell=bash
 TERMUX_SUBPKG_DESCRIPTION="The Glasgow Haskell Compiler"
 TERMUX_SUBPKG_DEPENDS="binutils-is-llvm | binutils, llvm, clang"
 
-TERMUX_SUBPKG_INCLUDE="lib/ghc-${TERMUX_PKG_VERSION}/ghc-${TERMUX_PKG_VERSION}"
+TERMUX_SUBPKG_INCLUDE="lib/ghc-$TERMUX_PKG_VERSION/ghc-$TERMUX_PKG_VERSION"
 
-for f in $(find ${TERMUX_PREFIX}/lib/ghc-${TERMUX_PKG_VERSION}/bin -type f -not -name "ghc-pkg*" -print); do
-	TERMUX_SUBPKG_INCLUDE+=" ${f/${TERMUX_PREFIX}\//}"
-done
-for f in $(find ${TERMUX_PREFIX}/bin -type f -not -name "ghc-pkg*" -print); do
-	TERMUX_SUBPKG_INCLUDE+=" ${f/${TERMUX_PREFIX}\//}"
-done
+while read -r file; do
+	TERMUX_SUBPKG_INCLUDE+=" ${file/$TERMUX_PREFIX\//}"
+done < <(find "$TERMUX_PREFIX"/lib/ghc-"$TERMUX_PKG_VERSION"/bin -type f -not -name "ghc-pkg*" -print)
+
+while read -r file; do
+	TERMUX_SUBPKG_INCLUDE+=" ${file/$TERMUX_PREFIX\//}"
+done < <(find "$TERMUX_PREFIX"/bin -type f -not -name "ghc-pkg*" -print)

--- a/packages/ghc-libs/ghc.subpackage.sh
+++ b/packages/ghc-libs/ghc.subpackage.sh
@@ -6,8 +6,8 @@ TERMUX_SUBPKG_INCLUDE="lib/ghc-$TERMUX_PKG_VERSION/ghc-$TERMUX_PKG_VERSION"
 
 while read -r file; do
 	TERMUX_SUBPKG_INCLUDE+=" ${file/$TERMUX_PREFIX\//}"
-done < <(find "$TERMUX_PREFIX"/lib/ghc-"$TERMUX_PKG_VERSION"/bin -type f -not -name "ghc-pkg*" -print)
+done < <(find "$TERMUX_PREFIX"/lib/ghc-"$TERMUX_PKG_VERSION"/bin -type f -or -type l -not -name "ghc-pkg*" -print)
 
 while read -r file; do
 	TERMUX_SUBPKG_INCLUDE+=" ${file/$TERMUX_PREFIX\//}"
-done < <(find "$TERMUX_PREFIX"/bin -type f -not -name "ghc-pkg*" -print)
+done < <(find "$TERMUX_PREFIX"/bin -type f -or -type l -not -name "ghc-pkg*" -print)

--- a/packages/ghc-libs/set-default-prefix.patch
+++ b/packages/ghc-libs/set-default-prefix.patch
@@ -1,5 +1,5 @@
---- ghc-8.10.7/libraries/Cabal/Cabal/Distribution/Simple/InstallDirs.hs	2021-05-25 20:11:34.000000000 +0530
-+++ ghc-8.10.7-patch/libraries/Cabal/Cabal/Distribution/Simple/InstallDirs.hs	2022-02-03 22:24:07.081585076 +0530
+--- ghc-8.10.7/libraries/Cabal/Cabal/src/Distribution/Simple/InstallDirs.hs	2021-05-25 20:11:34.000000000 +0530
++++ ghc-8.10.7-patch/libraries/Cabal/Cabal/src/Distribution/Simple/InstallDirs.hs	2022-02-03 22:24:07.081585076 +0530
 @@ -193,7 +193,7 @@
        else case buildOS of
             Windows -> do windowsProgramFilesDir <- getWindowsProgramFilesDir

--- a/packages/ghc-libs/undefined_symbols.patch
+++ b/packages/ghc-libs/undefined_symbols.patch
@@ -3,17 +3,17 @@
 @@ -12,6 +12,10 @@
  // internal headers
  #include "Trace.h"
- 
+
 +int TRACE_sched;
 +int TRACE_gc;
 +int TRACE_spark_sampled;
 +
  #if defined(TRACING)
- 
+
  #include "GetTime.h"
-@@ -28,10 +32,7 @@
+@@ -29,10 +29,7 @@
  #endif
- 
+
  // events
 -int TRACE_sched;
 -int TRACE_gc;

--- a/scripts/build/setup/termux_setup_cabal.sh
+++ b/scripts/build/setup/termux_setup_cabal.sh
@@ -1,7 +1,7 @@
 # shellcheck shell=bash
 termux_setup_cabal() {
 	if [[ "${TERMUX_ON_DEVICE_BUILD}" == "false" ]]; then
-		local TERMUX_CABAL_VERSION=3.6.2.0
+		local TERMUX_CABAL_VERSION=3.8.1.0
 		local TERMUX_CABAL_TAR="${TERMUX_COMMON_CACHEDIR}/cabal-${TERMUX_CABAL_VERSION}.tar.xz"
 
 		local TERMUX_CABAL_RUNTIME_FOLDER
@@ -18,7 +18,7 @@ termux_setup_cabal() {
 
 		termux_download "https://github.com/MrAdityaAlok/ghc-cross-tools/releases/download/cabal-install-v${TERMUX_CABAL_VERSION}/cabal-install-${TERMUX_CABAL_VERSION}.tar.xz" \
 			"${TERMUX_CABAL_TAR}" \
-			"f433e99cb3ff85239bd633f2ae2a370bfb7103f9db80e38199e0fda27897bdfe"
+			6a8c3c04414cbf1c805d2c1c63a09bdc879a7babd9ac1143370bdb6f6bf5a927
 
 		mkdir -p "${TERMUX_CABAL_RUNTIME_FOLDER}"
 		tar xf "${TERMUX_CABAL_TAR}" -C "${TERMUX_CABAL_RUNTIME_FOLDER}"

--- a/scripts/build/setup/termux_setup_ghc_cross_compiler.sh
+++ b/scripts/build/setup/termux_setup_ghc_cross_compiler.sh
@@ -55,7 +55,7 @@ __termux_haskell_setup_build_script() {
 
 # Utility function to setup a GHC cross-compiler toolchain targeting Android.
 termux_setup_ghc_cross_compiler() {
-	local TERMUX_GHC_VERSION="8.10.7"
+	local TERMUX_GHC_VERSION=9.2.5
 	local GHC_PREFIX="ghc-cross-${TERMUX_GHC_VERSION}-${TERMUX_ARCH}"
 	if [[ "${TERMUX_ON_DEVICE_BUILD}" == "false" ]]; then
 		local TERMUX_GHC_RUNTIME_FOLDER
@@ -83,10 +83,10 @@ termux_setup_ghc_cross_compiler() {
 		local CHECKSUMS
 		CHECKSUMS="$(
 			cat <<-EOF
-				aarch64:0912e8c6a8f4b362198c26129bb55f8e76edfcfbf38bfaf8b025a46429e6a887
-				arm:4f9acf98ee44eaebec6bce915507a934d1f880dd4c7ee679c075644e3bc41f78
-				i686:289a04baa67b8cbef401aebf8f5ffef90735e5a5b6e00ce39a50b82c134fe51b
-				x86_64:b43b4c8b80210c2b17ad4547d6d007163052edbd662495e0010b1c9b17d4f865
+				aarch64:47893a77abd35ce5f884bf9c67f8f0437dbcb297d5939e17a3ce7aa74c7d34b8
+				arm:dca3aa7a523054e5b472793afb0d750162052ffa762122c1200e5d832187bb86
+				i686:428c26a4c2a26737a9c031dbe7545c6514d9042cb28d926ffa8702c2930326c5
+				x86_64:1b27fa3dfa02cc9959b43a82b2881b55a1def397da7e7f7ff64406c666763f50
 			EOF
 		)"
 

